chips/sam4l/src/adc.rs

```diff
@@ -273,32 +293,61 @@ impl hil::adc::Adc for Adc {
                 // as the CPU clock
                 pm::enable_clock(Clock::PBA(PBAClock::ADCIFE));
                 nvic::enable(nvic::NvicIdx::ADCIFE);
-
-                // turn on GCLK10. The Generic Clock is needed for the ADC to
-                // work, although I have no idea what it is needed for. Its
-                // speed does not seem to matter, so let's just peg it to the
-                // CPU clock
-                scif::generic_clock_enable(scif::GenericClock::GCLK10, scif::ClockSource::CLK_CPU);
-
-                // determine clock divider
-                // we need the ADC_CLK to be a maximum of 1.5 MHz in frequency,
-                // so we need to find the PRESCAL value that will make this
-                // happen.
-                // Formula: f(ADC_CLK) = f(CLK_CPU)/2^(N+2) <= 1.5 MHz
-                // and we solve for N
-                // becomes: N <= ceil(log_2(f(CLK_CPU)/1500000)) - 2
-                let cpu_frequency = pm::get_system_frequency();
-                let divisor = (cpu_frequency + (1500000 - 1)) / 1500000; // ceiling of division
-                clock_divisor = math::log_base_two(math::closest_power_of_two(divisor)) - 2;
-                clock_divisor = cmp::min(cmp::max(clock_divisor, 0), 7); // keep in bounds
-                self.adc_clk_freq.set(cpu_frequency / (1 << (clock_divisor + 2)));
+                // the maximum sampling frequency with the RC clocks is 1/32th of their clock
+                // frequency. This is because of the minimum PRESCAL by a factor of 4 and the
+                // 7+1 cycles needed for conversion in continuous mode. Hence, 4*(7+1)=32.
+                if frequency <= 115000 / 32 {
```

Comment:
Okay, I'm pretty sure my prior math was a little wrong. (edited to note that)

So for RC32K, the expected range is 20 kHz to 44 kHz. You divide it by 32, to get a clock that ranges from 625 Hz to 1375 Hz, but you've got to assume the nominal of 1024 Hz. So if you want to sample once per second, you set the timer to fire every 1024 cycles (or divide the clock by an additional 2^10 in your case). So now you've got a range of 0.610 Hz to 1.34 Hz, when you expected 1 Hz exactly, which means your one second sampling period takes between 1.639 seconds (1/0.610) and 0.746 seconds (1/1.34). In the end, these numbers are calculated by nominal frequency / actual frequency. This is plus 64% to minus 25% error, regardless of sampling rate. (Sampling at 10 Hz would be 1/10 of the absolute error, but still 64% error)

So if we extend this to the other clocks, RCSYS is 1 - (113600 / 116000) = 2.07% fast to (113600/110000) - 1 = 3.27% slow.

The external oscillator which drives the CPU clock is +/- 30 ppm and therefore ranges from 47998560 to 48001440 Hz, which is +/- 0.003% error (+/- 30 per 1000000, which makes sense).

So takeaway, I would say that using RCSYS for slower applications is probably fine (3% error feels acceptable to me) as long as we clearly document what the expected error is. However, I would say that using RC32K is not acceptable.

Now, we don't actually have documentation on what the distribution of this error is. Perhaps the RC32K is a very narrow distribution of frequencies with a long tail, but there's always a chance that the error will be that bad (with the system still in spec).

@petarpenkov Does this make sense to you? I could still be making a mistake somewhere.